File: web/src/bridge/issueSelectionFingerprint.ts
import { sha256HexUtf8 } from "./selectionFingerprint";

function normalizeQualityIssueId(issueId: string): string | null {
  const trimmed = issueId.trim();
  if (!trimmed) return null;
  if (trimmed.startsWith("quality:")) {
    const payload = trimmed.slice("quality:".length);
    if (!payload || payload.startsWith("quality:")) return null;
    return `quality:${payload}`;
  }
  return `quality:${trimmed}`;
}

export function normalizeRepairIssueIds(issueIds: string[]): string[] {
  const seen = new Set<string>();
  const normalized: string[] = [];
  for (const raw of issueIds) {
    const resolved = normalizeQualityIssueId(raw);
    if (!resolved || seen.has(resolved)) continue;
    seen.add(resolved);
    normalized.push(resolved);
  }
  normalized.sort();
  return normalized;
}

export function issueSelectionFingerprint(issueIds: string[]): string {
  const normalized = normalizeRepairIssueIds(issueIds);
  return sha256HexUtf8(JSON.stringify({ issueIds: normalized }));
}

File: src/application/issue_selection_fingerprint.py
"""SHA-256 fingerprint of normalized quality repair issue id selection."""

from __future__ import annotations

import hashlib
import json

from application.quality_issue_detail import normalize_quality_issue_id

MAX_REPAIR_BATCH = 10


def normalize_repair_issue_ids(issue_ids: list[str]) -> list[str]:
    normalized: list[str] = []
    seen: set[str] = set()
    for raw in issue_ids:
        if not isinstance(raw, str):
            continue
        resolved = normalize_quality_issue_id(raw)
        if resolved is None or resolved in seen:
            continue
        seen.add(resolved)
        normalized.append(resolved)
    normalized.sort()
    return normalized


def issue_selection_fingerprint(issue_ids: list[str]) -> str:
    normalized = normalize_repair_issue_ids(issue_ids)
    payload = json.dumps({"issueIds": normalized}, sort_keys=True, separators=(",", ":"))
    return hashlib.sha256(payload.encode("utf-8")).hexdigest()
